### PR TITLE
Definition of network interface

### LINKED ISSUE #24: Define network interface
There's not need to implement any of the functions, but we must define the needed types for them and leave the functions defined as:

```
functionName :: someType
functionName = undefined
```

The currently expected functions are: 

For the users:

```
queryLeader
getSeathConfig
sendActionToLeader
sendSignedTxToLeader
sendRejectionToLeader
waitForActionConfirmation
```

In the leader:

```
sendChainToUsers
waitForChainSignatures
rebuildChainFrom
submitChain
sendConfirmationToUsers
```

### PR DESCRIPTION
**Pull request by Luis-omega**
closes #24

### CODE REVIEW

**Comment by Luis-omega** [comment]
Some open questions: 

## How to relate  `CoreConfiguration` with `NetworkConfiguration` ? 

We can put one inside the other or write a type like 
```haskell
data SeathConfiguration a b c d e f = SeathConfiguration {core::CoreConfiguration a b c d e f, network,NetworkConfiguration}
``` 
and then use it as base for the general `SeathMonad` or we can restrict the SeathMonad to just network or delete it an just pass around `NetworkConfiguration` (but we are boing to update it a lot in seath.

## We need to include a Leader query? 

## Shall we add the IP of the user to the UserAction to simplify the work of storing it?

**Comment by Luis-omega** [comment]
The latest are related to [this comment](https://github.com/mlabs-haskell/seath/issues/24#issuecomment-1478959401)

Also, part of the intention is to:
- have a variable number of participants in plutip test.
- Define a replacement for `Seath.Test.Examples.Addition.ContractSeath(mainTest)` that uses the new interface.
- Tried to avoid changes that conflict with the branch `misha/demo`

**Comment by Luis-omega** [comment]
Some changes that I still want to do: 
- Change the name of `Seath.Core.HandleActions` to something like `ActionsProcessing` or `ChainBuild`
- Define how to use mix `CTL` and `Quickcheck` (most probably getting rid of `Quickcheck` or running it and lift the result to `Contract`, but we won't be able to run property testing, that can be a problem to solve later in stage 3)

**Comment by Luis-omega** [comment]
Remaining work : 
- Solve all TODOs
- fully write `Seath.Examples.Addition.ContractSeath.newMainTest`
- Remove `Seath.Examples.Addition.SeathSetup`

**Comment by Luis-omega** [comment]
Something I still don't know how to do is the interaction between the seath network part and the CTL part.

Using a `SeathMonad` we can have

```haskell
-- env = LeaderNode | UserNode
newtype SeathMonad env a = ReaderT env Aff a
```

And then move between them by converthing `SeathMonad` and `Contract` to `Aff` and then use `liftAff`

Another option is to:


```haskell
newtype SeathMonad env a = ReaderT env Contract a
```
Then we can drop some parameters of the configuration for the network. A problem with this is that we would misuse `Contract` as we principally would want it

File: off-chain/src/Network/Types.purs [hunk 1]
```diff
@@ -1,155 +1,207 @@
...
+    -- | by the leader but aren't in processes right now.
+    -- TODO : put the right types in the following three
+    -- Maybe is : `MutableReference (OrderedMap (SomeUniqueID a) UserAction a)
+    actionsSent :: Array a
+  -- | `transactionsSent` is intended for both the `action` and it's
+  -- | corresponding `SignedTransaction` already send to the server.
+  , transactionsSent :: Array a
+  -- | Those signed transactions are confirmed to be in the chain 
+  -- | ready for submission.
+  -- | Once a transaction is confirmed to be done, we can safely
```
Comment (mikekeke) [inline off-chain/src/Network/Types.purs:182]: > Once a transaction is confirmed to be done

What "done" means here? Submitted to the node by the leader (via whatever submits it to the node in CLT) ?

File: off-chain/src/Network/Types.purs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+type Time = Int
+type MutableInt = Int
+
+-- This represent a Map that can remember the order of insertion.
+-- it seems that we can use purescript-ordered-collections with a 
+-- type like this.
+-- I would like to use purescript-ordered-collections, but the 
+-- Ord instance it has is over the keys, and then we would need 
+-- to cast it to a plain map quite often.
+type OrderedMap keys values =
```
Comment (mikekeke) [inline off-chain/src/Network/Types.purs:33]: > but the Ord instance it has is over the keys, and then we would need to cast it to a plain map quite often

Here is some [Haskell implementation](https://hackage.haskell.org/package/ordered-containers-0.2.3/docs/src/Data.Map.Ordered.Internal.html#OMap) I've found. Seems like it sacrifices some space, but avoids conversions.

File: off-chain/src/Network/Types.purs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+type MiliSeconds = Int
+type Time = Int
+type MutableInt = Int
+
+-- This represent a Map that can remember the order of insertion.
+-- it seems that we can use purescript-ordered-collections with a 
+-- type like this.
+-- I would like to use purescript-ordered-collections, but the 
+-- Ord instance it has is over the keys, and then we would need 
+-- to cast it to a plain map quite often.
```
Comment (mikekeke) [inline off-chain/src/Network/Types.purs:32]: > but the Ord instance it has is over the keys, and then we would need to cast it to a plain map quite often

Here is some [Haskell implementation](https://hackage.haskell.org/package/ordered-containers-0.2.3/docs/src/Data.Map.Ordered.Internal.html#OMap) I've found. Seems like it sacrifices some space, but avoids conversions.

**Pull request merged**